horovod/torch/__init__.py

```diff
@@ -63,6 +65,9 @@ def __init__(self, params, named_parameters, compression):
         if size() > 1:
             self._register_hooks()
 
+    def ignore_gradients(self, state):
```

Comment:
I'll think a bit more about a better name for this method.

Reply:
@alsrgv What about ``disable/enable_gradient_updates``

Reply:
Thanks for putting this together @andfoy!

This is a very useful capability, but personally, I would rather not add this method to the optimizer API.  I think it forces the user to make a lot of modifications on their end to experiment with local gradient accumulation, when it could be simplified instead to a single parameter to the DistributedOptimizer constructor:

```
opt = hvd.DistributedOptimizer(opt, aggregation_delay=5)
```

We could then manage all of the state internally without having to expose the optimizer API as a state machine, which I think would be more in line with our goals of making Horovod easy to plug into existing single-GPU scripts.  The `step()` function would then be a no-op in the case that we're currently aggregating locally.  

We lose some versatility, but I think it would greatly simplify things and make the feature less intimidating for people who want to try it out without modifying their training script (and less prone to error).

@alsrgv thoughts?

Reply:
Assuming the `aggregation_delay` is static, this sounds like a neat idea!

Reply:
Thanks for your suggestions, it sounds good, but I have a corner case (Like in our research), what would happen if the  aggregation_delay is dynamic and not static. e.g., In variable-sized batches

Reply:
Good question.  I was wondering if it would suffice to have a `opt.set_aggregation_delay(delay)` method?  That way advanced users could set it dynamically based on some criteria, and beginning users could just set it via the constructor.  Would that address the corner case you had in mind?

Reply:
@tgaddair, that sounds perfect! So, let me introduce the changes on the PR, and discuss ir further

Reply:
@andfoy, if the delay is dynamic, I'm wondering about a couple of things:
1. Will the number of synchronizations be the same across all workers?
2. Do you always know the aggregation delay before you start aggregating?

If not, we may still want to expose finer-grained controls.

Reply:
@alsrgv Addressing your two questions:
1. In my specific case, there should be a single synchronization step across all workers once they have finished their individual gradient accumulations.
2. Before starting an accumulation loop, the number of steps is known by each worker.

The general idea would be to synchronize once all workers have finished, which is limited by the worker that has the largest gradient accumulation iteration number. I don't know in which cases, the more general scenario (Unknown number of iterations, different synchronization scheme) would arise, but I think that in that case, it would be preferable to disable/enable gradient synchronization as it is needed.